File: anvil-common/src/main/java/org/anvilpowered/anvil/api/ServiceManagerImpl.java
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020 Cableguy20
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.anvil.api;

import com.google.common.reflect.TypeToken;
import com.google.inject.Injector;
import org.anvilpowered.anvil.api.misc.BindingExtensions;
import org.anvilpowered.anvil.common.misc.CommonBindingExtensions;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;
import java.util.function.Supplier;

@SuppressWarnings({"unchecked", "unused", "UnstableApiUsage"})
class ServiceManagerImpl implements ServiceManager {

    private final Map<TypeToken<?>, Supplier<?>> supplierBindings;
    private final Map<TypeToken<?>, Function<?, ?>> functionBindings;
    private Injector injector;
    static final EnvironmentManagerImpl environmentManager
        = new EnvironmentManagerImpl();

    public ServiceManagerImpl() {
        supplierBindings = new HashMap<>();
        functionBindings = new HashMap<>();
        registerBinding(EnvironmentManager.class, () -> environmentManager);
        registerBinding(Environment.Builder.class, EnvironmentBuilderImpl::new);
        registerBinding(BindingExtensions.class, CommonBindingExtensions::new);
    }

    void setInjector(Injector injector) {
        this.injector = injector;
    }

    @Override
    public <T> Supplier<T> provideSupplier(TypeToken<T> typeToken) {
        Supplier<T> supplier = (Supplier<T>) supplierBindings.get(typeToken);
        if (supplier != null) {
            return supplier;
        }
        return () -> injector.getInstance(BindingExtensions.getKey(typeToken));
    }

    @Override
    public <T> Supplier<T> provideSupplier(Class<T> clazz) {
        return provideSupplier(TypeToken.of(clazz));
    }

    @Override
    public <T> Supplier<T> provideSupplier(String name) {
        Supplier<T>[] suppliers = new Supplier[1];
        for (Map.Entry<TypeToken<?>, Supplier<?>> entry : supplierBindings.entrySet()) {
            if (entry.getKey().getRawType().getName().equalsIgnoreCase(name)) {
                suppliers[0] = (Supplier<T>) entry.getValue();
                break;
            }
        }
        if (suppliers[0] != null) {
            return suppliers[0];
        }
        return () -> Environment.getInstance(name, injector);
    }

    @Override
    public <T, R> Function<T, R> provideFunction(TypeToken<R> typeToken) {
        Function<T, R> function = (Function<T, R>) functionBindings.get(typeToken);
        if (function != null) {
            return function;
        }
        return injector -> ((Injector) injector)
            .getInstance(BindingExtensions.getKey(typeToken));
    }

    @Override
    public <T, R> Function<T, R> provideFunction(Class<R> clazz) {
        return provideFunction(TypeToken.of(clazz));
    }

    @Override
    public <T, R> Function<T, R> provideFunction(String name) {
        Function<T, R>[] functions = new Function[1];
        for (Map.Entry<TypeToken<?>, Function<?, ?>> entry : functionBindings.entrySet()) {
            if (entry.getKey().getRawType().getName().equalsIgnoreCase(name)) {
                functions[0] = (Function<T, R>) entry.getValue();
                break;
            }
        }
        if (functions[0] != null) {
            return functions[0];
        }
        return injector -> Environment.getInstance(name, (Injector) injector);
    }

    @Override
    public <T> T provide(TypeToken<T> typeToken) {
        return provideSupplier(typeToken).get();
    }

    @Override
    public <T> T provide(Class<T> clazz) {
        return provideSupplier(clazz).get();
    }

    @Override
    public <T> T provide(String name) {
        return this.<T>provideSupplier(name).get();
    }

    @Override
    public <T, R> R provide(TypeToken<R> typeToken, T input) {
        return provideFunction(typeToken).apply(input);
    }

    @Override
    public <T, R> R provide(Class<R> clazz, T input) {
        return provideFunction(clazz).apply(input);
    }

    @Override
    public <T, R> R provide(String name, T input) {
        return this.<T, R>provideFunction(name).apply(input);
    }

    @Override
    public <T> void registerBinding(TypeToken<T> typeToken, Supplier<T> supplier) {
        supplierBindings.put(typeToken, supplier);
    }

    @Override
    public <T> void registerBinding(Class<T> clazz, Supplier<T> supplier) {
        supplierBindings.put(TypeToken.of(clazz), supplier);
    }

    @Override
    public <T, R> void registerBinding(TypeToken<R> typeToken, Function<T, R> function) {
        functionBindings.put(typeToken, function);
    }

    @Override
    public <T, R> void registerBinding(Class<R> clazz, Function<T, R> function) {
        functionBindings.put(TypeToken.of(clazz), function);
    }
}

File: core/src/main/kotlin/org/anvilpowered/anvil/api/ServiceManagerImpl.kt
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020-2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.anvilpowered.anvil.api

import com.google.common.reflect.TypeToken
import com.google.inject.Injector
import org.anvilpowered.anvil.api.environment.Environment
import org.anvilpowered.anvil.api.environment.EnvironmentBuilderImpl
import org.anvilpowered.anvil.api.environment.EnvironmentManager
import org.anvilpowered.anvil.api.environment.EnvironmentManagerImpl
import java.util.function.Function
import java.util.function.Supplier

class ServiceManagerImpl : ServiceManager {
    private val supplierBindings: MutableMap<TypeToken<*>, Supplier<*>> = HashMap()
    private val functionBindings: MutableMap<TypeToken<*>, Function<*, *>> = HashMap()
    private lateinit var injector: Injector
    fun setInjector(injector: Injector) {
        this.injector = injector
    }

    override fun <T> provideSupplier(typeToken: TypeToken<T>): Supplier<T> {
        return supplierBindings[typeToken] as Supplier<T>
    }

    override fun <T> provideSupplier(clazz: Class<T>): Supplier<T> {
        return provideSupplier(TypeToken.of(clazz))
    }

    override fun <T> provideSupplier(name: String): Supplier<T> {
        val suppliers: Array<Supplier<T>> = arrayOf()
        for ((key, value) in supplierBindings) {
            if (key.rawType.name.equals(name, ignoreCase = true)) {
                suppliers[0] = value as Supplier<T>
                break
            }
        }
        return suppliers[0]
    }

    override fun <T, R> provideFunction(typeToken: TypeToken<R>): Function<T, R> {
        return functionBindings[typeToken] as Function<T, R>
    }

    override fun <T, R> provideFunction(clazz: Class<R>): Function<T, R> {
        return provideFunction(TypeToken.of(clazz))
    }

    override fun <T, R> provideFunction(name: String): Function<T, R> {
        val functions: Array<Function<T, R>> = arrayOf()
        for ((key, value) in functionBindings) {
            if (key.rawType.name.equals(name, ignoreCase = true)) {
                functions[0] = value as Function<T, R>
                break
            }
        }
        return functions[0]
    }

    override fun <T> provide(typeToken: TypeToken<T>): T {
        return provideSupplier(typeToken).get()
    }

    override fun <T> provide(clazz: Class<T>): T {
        return provideSupplier(clazz).get()
    }

    override fun <T> provide(name: String): T {
        return this.provideSupplier<T>(name).get()
    }

    override fun <T, R> provide(typeToken: TypeToken<R>, input: T): R {
        return provideFunction<Any, R>(typeToken).apply(input!!)
    }

    override fun <T, R> provide(clazz: Class<R>, input: T): R {
        return provideFunction<Any, R>(clazz).apply(input!!)
    }

    override fun <T, R> provide(name: String, input: T): R {
        return this.provideFunction<T, R>(name).apply(input)
    }

    override fun <T> registerBinding(typeToken: TypeToken<T>, supplier: Supplier<T>) {
        supplierBindings[typeToken] = supplier
    }

    override fun <T> registerBinding(clazz: Class<T>, supplier: Supplier<T>) {
        supplierBindings[TypeToken.of(clazz)] = supplier
    }

    override fun <T, R> registerBinding(typeToken: TypeToken<R>, function: Function<T, R>) {
        functionBindings[typeToken] = function
    }

    override fun <T, R> registerBinding(clazz: Class<R>, function: Function<T, R>) {
        functionBindings[TypeToken.of(clazz)] = function
    }

    companion object {
        val environmentManager = EnvironmentManagerImpl()
    }

    init {
        registerBinding<EnvironmentManager>(EnvironmentManager::class.java) { environmentManager }
        registerBinding<Environment.Builder>(Environment.Builder::class.java) { EnvironmentBuilderImpl() }
    }
}